worker-message-prioritization-distribution/src/main/java/com/github/workerframework/workermessageprioritization/redistribution/shovel/ShovelStateChecker.java

```diff
@@ -0,0 +1,112 @@
+                final Instant shovelCreationTimeUTC = shovelNameToCreationTimeUTC.computeIfAbsent(shovelName, s -> Instant.now());
+                final long shovelCreationTimeUTCMilliseconds = shovelCreationTimeUTC.toEpochMilli();
+
+                final Instant timeNowUTC = Instant.now();
+                final long timeNowUTCMilliseconds = timeNowUTC.toEpochMilli();
+
+                if (timeNowUTCMilliseconds - shovelCreationTimeUTCMilliseconds >= nonRunningShovelTimeoutMilliseconds) {   
+                    LOGGER.error("Shovel named {} was created at {}. The time now is {}. "
+                        + "The shovel is not yet in a 'running' state. It's current state is '{}'. "
+                        + "As the non-running shovel timeout of {} milliseconds has been reached, the shovel is now going to be deleted. "
+                        + "Please check the RabbitMQ logs for more details. "
+                        + "Shovel creation will be attempted later if the shovel is still required.",
+                                 shovelName,
+                                 shovelCreationTimeUTC,
+                                 timeNowUTC,
+                                 retrievedShovel.getState().toString().toLowerCase(),
+                                 nonRunningShovelTimeoutMilliseconds);                    
+ 
+                    shovelsApi.getApi().delete(rabbitMQVHost, shovelName);
+
+                    shovelNameToCreationTimeUTC.remove(shovelName);
+                } else {
+                    LOGGER.debug("Shovel named {} was created at {}. The time now is {}. "
+                        + "The shovel is not yet in a 'running' state. It's current state is '{}'. "
+                        + "However, as the non-running shovel timeout of {} milliseconds has not been reached yet, "
+                        + "the shovel will not be deleted at this time.",
+                                 shovelName,
+                                 shovelCreationTimeUTC,
+                                 timeNowUTC,
+                                 retrievedShovel.getState().toString().toLowerCase(),
+                                 nonRunningShovelTimeoutMilliseconds);
+                }
+            } else {
+                // Else the shovel is running, in which case we can remove it from the map as we don't need to keep it's creation time
+                shovelNameToCreationTimeUTC.remove(shovelName);
+            }
+        }
+
+        // Remove any shovels that have been deleted normally (e.g. automatically by RabbitMQ via src-delete-after) from the map
+        final List<String> retrievedShovelNames = retrievedShovels
+            .stream()
+            .map(RetrievedShovel::getName)
+            .collect(Collectors.toList());
+
+        shovelNameToCreationTimeUTC.entrySet().removeIf(entry -> !retrievedShovelNames.contains(entry.getKey()));
```

Comment:
@frankmccarry  There's various places here where we need to remove entries from this map (lines 86, 100, and 110). That's assuming I haven't forgotten any more scenarios.

I wonder would it be simpler to just have the map expire entries automatically like I had [here](https://github.com/WorkerFramework/worker-message-prioritization/pull/4/commits/aef03260a766b15c946c6ed7b127efaa61e426b5), e.g. after 12 or 24 hours? And remove all the calls to delete from this class?

Another thing to note is that this map is shared between 2 threads, A way to possibly simplify it would be for this class to own the map, and be responsible for inserting the timestamp into it when it runs, so, rather than recording shovel creation time like we do now, we could instead record 'time shovel has been observed in non-running state'?

Reply:
So, doing both these things might look something like:

```
package com.github.workerframework.workermessageprioritization.redistribution.shovel;

import com.github.workerframework.workermessageprioritization.rabbitmq.RabbitManagementApi;
import com.github.workerframework.workermessageprioritization.rabbitmq.RetrievedShovel;
import com.github.workerframework.workermessageprioritization.rabbitmq.ShovelState;
import com.github.workerframework.workermessageprioritization.rabbitmq.ShovelsApi;
import java.time.Instant;
import java.util.HashMap;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public final class ShovelStateChecker implements Runnable
{
    private static final Logger LOGGER = LoggerFactory.getLogger(ShovelStateChecker.class);

    private final RabbitManagementApi<ShovelsApi> shovelsApi;
    private final Map<String, Instant> shovelNameToTimeObservedInNonRunningState;
    private final String rabbitMQVHost;
    private final long nonRunningShovelTimeoutMilliseconds;

    public ShovelStateChecker(
        final RabbitManagementApi<ShovelsApi> shovelsApi,
        final Map<String, Instant> shovelNameToCreationTimeUTC,
        final String rabbitMQVHost,
        final long nonRunningShovelTimeoutMilliseconds)
    {
        this.shovelsApi = shovelsApi;
        this.shovelNameToTimeObservedInNonRunningState = ExpiringMap.builder()
            .expiration(12, TimeUnit.HOURS)
            .build();
        this.rabbitMQVHost = rabbitMQVHost;
        this.nonRunningShovelTimeoutMilliseconds = nonRunningShovelTimeoutMilliseconds;
    }

    @Override
    public void run()
    {        
        final List<RetrievedShovel> retrievedShovels = shovelsApi.getApi().getShovels();

        LOGGER.debug("Read the following list of shovels from the RabbitMQ API: {}", retrievedShovels);

        for (final RetrievedShovel retrievedShovel : retrievedShovels) {

            final String shovelName = retrievedShovel.getName();

            if (retrievedShovel.getState() != ShovelState.RUNNING) {

                final Instant timeObservedInNonRunningState = shovelNameToTimeObservedInNonRunningState
                    .computeIfAbsent(shovelName, s -> Instant.now());
                final long timeObservedInNonRunningStateMilliseconds = timeObservedInNonRunningState.toEpochMilli();

                final Instant timeNow = Instant.now();
                final long timeNowMilliseconds = timeNow.toEpochMilli();

                if (timeNowMilliseconds - timeObservedInNonRunningStateMilliseconds >= nonRunningShovelTimeoutMilliseconds) {   
                    LOGGER.error("Shovel named {} was observed in a non-running state at {}. The time now is {}. "
                        + "It's current state is '{}'. "
                        + "As the non-running shovel timeout of {} milliseconds has been reached, the shovel is now going to be deleted. "
                        + "Please check the RabbitMQ logs for more details. "
                        + "Shovel creation will be attempted later if the shovel is still required.",
                                 shovelName,
                                 timeObservedInNonRunningState,
                                 timeNow,
                                 retrievedShovel.getState().toString().toLowerCase(),
                                 nonRunningShovelTimeoutMilliseconds);                    
 
                    shovelsApi.getApi().delete(rabbitMQVHost, shovelName);
                } else {
                    LOGGER.debug("Shovel named {} was observed in a non-running state at {}. The time now is {}. "
                        + "The shovel is not yet in a 'running' state. It's current state is '{}'. "
                        + "However, as the non-running shovel timeout of {} milliseconds has not been reached yet, "
                        + "the shovel will not be deleted at this time.",
                                 shovelName,
                                 timeObservedInNonRunningState,
                                 timeNow,
                                 retrievedShovel.getState().toString().toLowerCase(),
                                 nonRunningShovelTimeoutMilliseconds);
                }
            } 
        }
    }
}

```

Reply:
I've pushed up those changes in this commit: https://github.com/WorkerFramework/worker-message-prioritization/pull/4/commits/f0341e424f9c1bf0fb0df47842fb77f6f9217178

I think its a bit cleaner with regards to map removal, and not sharing the map between threads, but I'm happy to revert it if you prefer the way it was before.